File: code-example-tests/java/utilities/comparison-library/src/main/java/mongodb/comparison/KernelBridge.java
package mongodb.comparison;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ArrayBlockingQueue;

/**
 * Bridge to the comparison kernel (a native Go binary committed under
 * tools/comparison-kernel/bin/). Delegates file-based comparisons to the
 * language-agnostic kernel for consistency across all driver test suites.
 *
 * <p>A pool of persistent kernel processes is maintained so that concurrent test
 * threads each get an exclusive connection without contention. Each connection reads
 * newline-delimited JSON requests from stdin and writes newline-delimited JSON
 * responses to stdout, amortising process startup cost across many calls.
 */
class KernelBridge {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    /** Wire-protocol version this bridge speaks. Must match the kernel's. */
    private static final int PROTOCOL_VERSION = 1;

    // Pool of reusable kernel connections — one per concurrent caller.
    // Sized to the host's processor count, with a floor of 4 (so single- and
    // dual-core CI runners still get useful parallelism) and a ceiling of 16
    // (each kernel process holds a few MB of resident memory, so an unbounded
    // pool on a many-core CI box would waste memory for diminishing returns).
    private static final int POOL_SIZE = Math.min(
            16, Math.max(4, Runtime.getRuntime().availableProcessors()));
    private static final ArrayBlockingQueue<KernelConnection> POOL =
            new ArrayBlockingQueue<>(POOL_SIZE);

    private static volatile Path cachedBinaryPath = null;

    /** A single live connection to one kernel process. */
    private static final class KernelConnection {
        final Process       process;
        final PrintWriter   writer;
        final BufferedReader reader;

        KernelConnection(Process process, PrintWriter writer, BufferedReader reader) {
            this.process = process;
            this.writer  = writer;
            this.reader  = reader;
        }

        boolean isAlive() { return process.isAlive(); }

        void destroy() {
            try { writer.close(); } catch (Exception ignored) {}
            process.destroy();
        }
    }

    static {
        // Drain and destroy all pooled connections on JVM shutdown.
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            List<KernelConnection> connections = new ArrayList<>();
            POOL.drainTo(connections);
            for (KernelConnection conn : connections) {
                conn.destroy();
            }
        }));
    }

    /**
     * Map the JVM's os.name / os.arch system properties onto Go's GOOS / GOARCH
     * naming used by tools/comparison-kernel/build.sh.
     */
    static String binaryName() {
        String osName = System.getProperty("os.name", "").toLowerCase();
        String osArch = System.getProperty("os.arch", "").toLowerCase();

        String goos;
        if (osName.contains("mac") || osName.contains("darwin")) goos = "darwin";
        else if (osName.contains("linux")) goos = "linux";
        else if (osName.contains("windows")) goos = "windows";
        else throw new UnsupportedOperationException(
                "comparison-kernel: unsupported host OS " + osName);

        String goarch;
        if (osArch.equals("aarch64") || osArch.equals("arm64")) goarch = "arm64";
        else if (osArch.equals("x86_64") || osArch.equals("amd64")) goarch = "amd64";
        else throw new UnsupportedOperationException(
                "comparison-kernel: unsupported host arch " + osArch);

        String suffix = goos.equals("windows") ? ".exe" : "";
        return "comparison-kernel-" + goos + "-" + goarch + suffix;
    }

    /**
     * Find the platform-specific kernel binary by walking up from the current
     * working directory.
     */
    static Path findBinaryPath() throws IOException {
        String name = binaryName();
        Path dir = Paths.get(System.getProperty("user.dir")).toAbsolutePath();
        for (int i = 0; i < 12; i++) {
            Path candidate = dir.resolve("tools/comparison-kernel/bin").resolve(name);
            if (Files.exists(candidate)) {
                return candidate;
            }
            Path parent = dir.getParent();
            if (parent == null || parent.equals(dir)) {
                break;
            }
            dir = parent;
        }
        throw new IOException(
                name + " not found; build it with: ./build.sh in "
                        + "code-example-tests/tools/comparison-kernel");
    }

    /**
     * Resolve a relative file path using the same strategies as ExpectedOutputParser.
     */
    static Path resolveFilePath(String filePath) {
        if (Paths.get(filePath).isAbsolute()) {
            return Paths.get(filePath);
        }
        Path cwd = Paths.get(System.getProperty("user.dir"));
        Path[] candidates = {
            cwd.resolve(filePath),
            cwd.resolve("examples").resolve(filePath),
            cwd.getParent() != null ? cwd.getParent().resolve("examples").resolve(filePath) : null,
            cwd.resolve("src/test/resources").resolve(filePath),
            cwd.resolve("src/test/resources/examples").resolve(filePath),
        };
        for (Path candidate : candidates) {
            if (candidate != null && Files.exists(candidate)) {
                return candidate;
            }
        }
        // Walk up to 3 parent levels looking for examples/
        Path searchDir = cwd;
        for (int i = 0; i < 3; i++) {
            Path p = searchDir.resolve("examples").resolve(filePath);
            if (Files.exists(p)) return p;
            searchDir = searchDir.getParent();
            if (searchDir == null) break;
        }
        return Paths.get(filePath);
    }

    /**
     * Normalize actual results to a JSON-serialisable List for the kernel's
     * "actual" wire field. The value is embedded directly in the request JSON,
     * not stringified, so the kernel doesn't have to double-decode it.
     */
    static List<Object> serializeActual(Object actual) {
        Object normalized = MongoTypeNormalizer.normalizeValue(actual);
        if (normalized instanceof List<?> list) {
            return new ArrayList<>(list);
        }
        if (normalized == null) {
            return List.of();
        }
        return List.of(normalized);
    }

    /**
     * Spawn a fresh kernel process and wrap it in a KernelConnection.
     */
    private static KernelConnection newConnection() throws IOException {
        // Find the platform-specific binary once; subsequent calls reuse the
        // cached path.
        if (cachedBinaryPath == null) {
            synchronized (KernelBridge.class) {
                if (cachedBinaryPath == null) {
                    cachedBinaryPath = findBinaryPath();
                }
            }
        }
        ProcessBuilder pb = new ProcessBuilder(cachedBinaryPath.toString());
        pb.redirectErrorStream(false);
        Process process = pb.start();
        // Drain stderr asynchronously so the pipe buffer never fills and deadlocks
        // the child process. Mirrors what the C# bridge does with Task.Run.
        Thread stderrDrainer = new Thread(() -> {
            try (BufferedReader err = new BufferedReader(
                    new InputStreamReader(process.getErrorStream(), StandardCharsets.UTF_8))) {
                String line;
                while ((line = err.readLine()) != null) {
                    System.err.println("[comparison-kernel] " + line);
                }
            } catch (IOException ignored) {
                // Stream closed when the process exits — nothing left to drain.
            }
        }, "kernel-stderr-drainer");
        stderrDrainer.setDaemon(true);
        stderrDrainer.start();
        PrintWriter writer = new PrintWriter(
                new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8),
                /* autoFlush= */ true);
        BufferedReader reader = new BufferedReader(
                new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8));
        return new KernelConnection(process, writer, reader);
    }

    /**
     * Borrow a connection from the pool, creating one if the pool is empty.
     */
    private static KernelConnection borrowConnection() throws IOException {
        KernelConnection conn = POOL.poll();
        if (conn != null && conn.isAlive()) {
            return conn;
        }
        if (conn != null) {
            conn.destroy(); // stale — discard and create a fresh one
        }
        return newConnection();
    }

    /**
     * Return a connection to the pool. Destroys it if the pool is already full
     * or the process has died.
     */
    private static void returnConnection(KernelConnection conn) {
        if (conn.isAlive() && POOL.offer(conn)) {
            return; // successfully returned
        }
        conn.destroy(); // pool full or process dead — clean up
    }

    /**
     * Run any comparison via the kernel process pool. The expected value is
     * serialised into raw text the kernel can parse (existing file path → file
     * contents; JSON-shaped string → as-is; everything else → BSON-normalised
     * and JSON-encoded). The actual value is normalised into a
     * JSON-serialisable list embedded directly in the request body.
     *
     * <p>Throws IOException if the kernel binary is missing or the round-trip
     * fails; callers should surface the error rather than masking it.
     */
    static ComparisonResult compare(
            Object expected, Object actual, ComparisonOptions options) throws IOException {
        return compare(expected, actual, options, null);
    }

    /**
     * Like {@link #compare(Object, Object, ComparisonOptions)} but also passes
     * a schema descriptor to the kernel for ShouldResemble-style validation.
     */
    static ComparisonResult compare(
            Object expected, Object actual, ComparisonOptions options, Schema schema)
            throws IOException {
        String expectedContent = serializeExpected(expected);
        List<Object> actualPayload = serializeActual(actual);

        Map<String, Object> optsMap = new LinkedHashMap<>();
        optsMap.put("comparisonType",
                options != null && options.comparisonType() != null
                        ? options.comparisonType().name().toLowerCase()
                        : "");
        optsMap.put("ignoreFieldValues",
                options != null && options.ignoreFieldValues() != null
                        ? options.ignoreFieldValues()
                        : List.of());
        if (schema != null) {
            Map<String, Object> schemaMap = new LinkedHashMap<>();
            schemaMap.put("count", schema.getCount());
            schemaMap.put("requiredFields",
                    schema.getRequiredFields() != null ? schema.getRequiredFields() : List.of());
            Map<String, Object> normalizedFV = new LinkedHashMap<>();
            for (Map.Entry<String, Object> e : schema.getFieldValues().entrySet()) {
                normalizedFV.put(e.getKey(), MongoTypeNormalizer.normalizeValue(e.getValue()));
            }
            schemaMap.put("fieldValues", normalizedFV);
            optsMap.put("schema", schemaMap);
        }

        Map<String, Object> request = new LinkedHashMap<>();
        request.put("protocolVersion", PROTOCOL_VERSION);
        request.put("expected", expectedContent);
        request.put("actual", actualPayload);
        request.put("options", optsMap);
        String requestLine = MAPPER.writeValueAsString(request);

        // Borrow an exclusive connection — no lock held during I/O.
        KernelConnection conn = borrowConnection();
        String responseLine;
        try {
            conn.writer.println(requestLine);
            if (conn.writer.checkError()) {
                conn.destroy();
                throw new IOException("Failed to write request to kernel process");
            }
            responseLine = conn.reader.readLine();
            returnConnection(conn);
        } catch (IOException e) {
            conn.destroy();
            throw e;
        }

        if (responseLine == null || responseLine.isBlank()) {
            throw new IOException("Kernel returned empty response");
        }

        @SuppressWarnings("unchecked")
        Map<String, Object> response = MAPPER.readValue(responseLine, Map.class);
        boolean isMatch = Boolean.TRUE.equals(response.get("isMatch"));
        if (isMatch) {
            return ComparisonResult.success();
        }

        @SuppressWarnings("unchecked")
        List<Map<String, Object>> errors = (List<Map<String, Object>>) response.get("errors");
        if (errors != null && !errors.isEmpty()) {
            List<ComparisonError> compErrors = errors.stream()
                    .map(e -> new ComparisonError(
                            String.valueOf(e.getOrDefault("path", "")),
                            String.valueOf(e.getOrDefault("expected", "")),
                            String.valueOf(e.getOrDefault("actual", "")),
                            String.valueOf(e.getOrDefault("message", "Comparison failed"))))
                    .toList();
            return ComparisonResult.failure(compErrors);
        }

        String errorMsg = (String) response.get("error");
        return ComparisonResult.failure(errorMsg != null ? errorMsg : "Comparison failed");
    }

    /**
     * Turn an expected value into raw text the kernel can parse:
     * <ul>
     *   <li>existing file path                → file contents</li>
     *   <li>string starting with "{" or "["  → as-is (MongoDB-syntax content)</li>
     *   <li>everything else                  → BSON-normalised and JSON-encoded</li>
     * </ul>
     */
    static String serializeExpected(Object expected) throws IOException {
        if (expected == null) {
            return "null";
        }
        if (expected instanceof Path pathExpected) {
            return Files.readString(pathExpected, StandardCharsets.UTF_8);
        }
        if (expected instanceof String s) {
            if (looksLikeFilePath(s)) {
                Path resolved = resolveFilePath(s);
                if (Files.exists(resolved)) {
                    return Files.readString(resolved, StandardCharsets.UTF_8);
                }
            }
            String trimmed = s.stripLeading();
            if (trimmed.startsWith("{") || trimmed.startsWith("[")) {
                return s;
            }
            return MAPPER.writeValueAsString(s);
        }
        return MAPPER.writeValueAsString(MongoTypeNormalizer.normalizeValue(expected));
    }

    /**
     * Heuristic: a string could plausibly be a file path if it is single-line,
     * shorter than 1 KiB, and doesn't start with a JSON marker.
     */
    private static boolean looksLikeFilePath(String s) {
        if (s == null || s.isEmpty()) return false;
        if (s.length() > 1024) return false;
        if (s.indexOf('\n') >= 0 || s.indexOf('\r') >= 0) return false;
        String trimmed = s.stripLeading();
        return !trimmed.startsWith("{") && !trimmed.startsWith("[");
    }
}

File: code-example-tests/python/pymongo/utils/comparison/kernel_bridge.py
"""
kernel_bridge.py — Synchronous bridge to the comparison-kernel binary.

The kernel is a native Go binary, one per supported platform, committed under
``tools/comparison-kernel/bin/``.  This bridge spawns the binary matching the
host OS/arch as a child process and communicates over the newline-delimited
JSON protocol::

    → one JSON request object per line written to the process's stdin
    ← one JSON response object per line read from the process's stdout

No runtime dependency is required — the binary runs directly.
"""

from __future__ import annotations

import json
import platform
import subprocess
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional

#: Wire-protocol version this bridge speaks. Must match the kernel's.
PROTOCOL_VERSION = 1


# Map Python's sys.platform / platform.machine() onto Go's GOOS / GOARCH naming.
_GOOS_MAP = {"darwin": "darwin", "linux": "linux", "win32": "windows"}
_GOARCH_MAP = {
    "arm64": "arm64",
    "aarch64": "arm64",
    "x86_64": "amd64",
    "amd64": "amd64",
}


def _resolve_kernel_binary() -> str:
    """Return the absolute path of the kernel binary for the current host."""
    goos = _GOOS_MAP.get(sys.platform)
    goarch = _GOARCH_MAP.get(platform.machine().lower())
    if not goos or not goarch:
        raise RuntimeError(
            f"comparison-kernel: unsupported host platform "
            f"{sys.platform}/{platform.machine()}. "
            "Add a target to tools/comparison-kernel/build.sh and rebuild."
        )
    suffix = ".exe" if goos == "windows" else ""
    binary_name = f"comparison-kernel-{goos}-{goarch}{suffix}"

    current = Path(__file__).resolve().parent
    for _ in range(12):
        candidate = current / "tools" / "comparison-kernel" / "bin" / binary_name
        if candidate.exists():
            return str(candidate)
        parent = current.parent
        if parent == current:
            break
        current = parent
    raise FileNotFoundError(
        f"{binary_name} not found. Build it with: ./build.sh in "
        "code-example-tests/tools/comparison-kernel"
    )


def _normalize_for_wire(value: Any) -> Any:
    """
    Recursively turn PyMongo/BSON types into JSON-serialisable equivalents
    the kernel knows how to consume:

      * ``datetime`` → ``{"$date": <iso>}`` (kernel collapses this to a string)
      * ``bson.Int64`` / ``bson.Int32`` → ``int(value)``
      * ``bson.ObjectId`` / ``bson.Decimal128`` / other bson scalars → ``str(value)``
      * ``bytes`` → hex string (so comparisons against expected hex literals work)
      * lists / dicts → walked recursively
      * everything else → passed through unchanged
    """
    from datetime import datetime

    if isinstance(value, datetime):
        return {"$date": value.isoformat()}
    if isinstance(value, bytes):
        return value.hex()
    if isinstance(value, list):
        return [_normalize_for_wire(v) for v in value]
    if isinstance(value, tuple):
        return [_normalize_for_wire(v) for v in value]
    if isinstance(value, dict):
        return {k: _normalize_for_wire(v) for k, v in value.items()}
    module = value.__class__.__module__
    if module.startswith("bson"):
        # BSON integer wrappers (Int32, Int64) carry numeric semantics; keeping
        # them as ints lets the kernel compare them against plain JSON numbers.
        cls_name = value.__class__.__name__
        if cls_name in ("Int64", "Int32") or isinstance(value, int) and not isinstance(value, bool):
            return int(value)
        return str(value)
    return value


def _serialize_actual(actual: Any) -> List[Any]:
    """
    Normalise *actual* into a JSON-serialisable list for the kernel's
    ``actual`` wire field. Single values are wrapped so the kernel always
    receives a document list.
    """
    if actual is None:
        return [None]
    normalized = _normalize_for_wire(actual)
    return normalized if isinstance(normalized, list) else [normalized]


def kernel_compare(
    expected_content: str,
    actual: Any,
    comparison_type: Optional[str] = None,
    ignore_field_values: Optional[List[str]] = None,
    schema: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """
    Send a synchronous comparison request to the kernel.

    Parameters
    ----------
    expected_content:
        Raw text of the expected output content.
    actual:
        Actual results from the test. Wrapped in a JSON array if necessary.
    comparison_type:
        ``"ordered"`` or ``None`` (kernel default: unordered).
    ignore_field_values:
        Field names whose values should be ignored during comparison.
    schema:
        Optional schema for ``shouldResemble``-style validation.

    Returns
    -------
    dict
        ``{"isMatch": bool, "errors": [...]}``
    """
    binary_path = _resolve_kernel_binary()
    actual_payload = _serialize_actual(actual)

    options: Dict[str, Any] = {}
    if comparison_type:
        options["comparisonType"] = comparison_type
    if ignore_field_values:
        options["ignoreFieldValues"] = ignore_field_values
    if schema is not None:
        options["schema"] = schema

    request = json.dumps(
        {
            "protocolVersion": PROTOCOL_VERSION,
            "expected": expected_content,
            "actual": actual_payload,
            "options": options,
        }
    )

    try:
        result = subprocess.run(
            [binary_path],
            input=request + "\n",
            capture_output=True,
            text=True,
            timeout=30,
        )
    except FileNotFoundError:
        raise RuntimeError(
            f"comparison-kernel binary not found at {binary_path}. "
            "Build it with: ./build.sh in code-example-tests/tools/comparison-kernel"
        )
    except subprocess.TimeoutExpired:
        raise RuntimeError("comparison-kernel timed out after 30 seconds")

    if result.returncode != 0:
        raise RuntimeError(
            f"comparison-kernel exited with code {result.returncode}: "
            f"{result.stderr.strip()}"
        )

    line = result.stdout.strip()
    if not line:
        raise RuntimeError("comparison-kernel returned empty response")

    return json.loads(line)
